kernel/src/pci.c

```diff
@@ -0,0 +1,93 @@
+#define DEVICE_TYPE_OFFSET 0x08
+
+#define DEVICE_CLASS_MASK 0xff000000
+#define DEVICE_SUBCLASS_MASK 0xffff0000
+#define DEVICE_PROG_IF_MASK 0xffffff00
+
+#define CONFIG_ADDRESS_ZERO 0x80000000
+
+// black-box, don't look!
+__attribute__((always_inline)) uint32_t create_config_address(uint8_t bus, uint8_t slot,
+                                                              uint8_t func, uint8_t offset) {
+    return (uint32_t)(((uint32_t)bus << 16) | ((uint32_t)(slot & 0x1F) << 11) |
+                      ((uint32_t)(func & 0x7) << 8) | ((uint32_t)offset & 0xfc) |
+                      ((uint32_t)0x80000000));
+}
+
+uint8_t pci_read_config_u8(uint32_t base_address, uint8_t offset) {
+    uint32_t config_address = (base_address & 0xffffff00) | (offset & ~0b11);
+    port_out_u32(PCI_REG_CONFIG_ADDRESS, config_address);
+    uint32_t config_data = port_in_u32(PCI_REG_CONFIG_DATA) >> ((offset & 0b11) * 8);
+    return (uint8_t)(config_data & 0xff);
+}
+
+uint16_t pci_read_config_u16(uint32_t base_address, uint8_t offset) {
+    uint32_t config_address = (base_address & 0xffffff00) | (offset & ~0b11);
+    port_out_u32(PCI_REG_CONFIG_ADDRESS, config_address);
+    uint32_t config_data = port_in_u32(PCI_REG_CONFIG_DATA) >> ((offset & 0b10) * 8);
+    return (uint16_t)(config_data & 0xffff);
+}
+
+uint32_t pci_read_config_u32(uint32_t base_address, uint8_t offset) {
+    uint32_t config_address = (base_address & 0xffffff00) | (offset & ~0b11);
+    port_out_u32(PCI_REG_CONFIG_ADDRESS, config_address);
+    uint32_t config_data = port_in_u32(PCI_REG_CONFIG_DATA);
+    return config_data;
+}
+
+// returns vendor ID, garuanteed to be 0xffff iff device dosen't exist.
+uint16_t get_vendor(uint32_t address) { return pci_read_config_u16(address, VENDOR_ID_OFFSET); }
+
+// returns true if device has multiple distinct functions
+bool device_is_multi_func(uint32_t address) {
+    return (pci_read_config_u8(address, HEADER_TYPE_OFFSET) & 0x80) != 0;
+}
+
+// Returns the next valid address
+uint32_t next_base_address(uint32_t last) {
+    if (last == 0) return CONFIG_ADDRESS_ZERO; // Special exception, return first address.
+    uint32_t bus = (last & 0x00ff0000) >> 16;
+    uint32_t slot = (last & 0x0000f800) >> 11;
+    uint32_t func = (last & 0x00000700) >> 8;
+
+    func++;
+    if (func < 8 && (func > 1 || ((get_vendor(last) != 0xffff) && device_is_multi_func(last)))) {
+        return create_config_address(bus, slot, func, 0);
+    }
+    func = 0;
+    slot++;
+    if (slot < 32) {
+        return create_config_address(bus, slot, func, 0);
+    }
+    slot = 0;
+    bus++;
+    if (bus < 256) {
+        return create_config_address(bus, slot, func, 0);
+    }
+    return 0; // No remaining addresses
+}
+
+uint32_t pci_find_next_device(uint32_t target_device_type, uint32_t mask, uint32_t start) {
+    uint32_t address = start;
+    do {
+        address = next_base_address(address);
+        if (get_vendor(address) != 0xffff) {
```

Comment:
get_vendor is used twice and if you look at: https://wiki.osdev.org/PCI#PCI_Device_Structure (Header type 0x00) you can see that there isn't much meaning in creating a specific macro in this case. It just compares the vendor id. I think that you there could instead be a define for NO_DEVICE_VENDOR_ID or something similar.

Reply:
since get_vendor is only ever used to check for non-existent devices it would probably be more readable to replace it with a device_exists function similar to device_is_multi_func.